planner/core/planbuilder.go

```diff
@@ -692,8 +693,17 @@ func getPhysicalIDs(tblInfo *model.TableInfo, partitionNames []model.CIStr) ([]i
 }
 
 func (b *PlanBuilder) buildAnalyzeTable(as *ast.AnalyzeTableStmt) (Plan, error) {
+	checker := privilege.GetPrivilegeManager(b.ctx)
 	p := &Analyze{MaxNumBuckets: as.MaxNumBuckets}
 	for _, tbl := range as.TableNames {
+		// MySQL will analyze each table and stop on the first to error without privs
+		user := b.ctx.GetSessionVars().User
+		if checker != nil && user != nil &&
+			!(checker.RequestVerification(tbl.Schema.O, tbl.Name.O, "", mysql.SelectPriv) &&
+				checker.RequestVerification(tbl.Schema.O, tbl.Name.O, "", mysql.InsertPriv)) {
+			return nil, ErrTableaccessDenied.GenWithStackByArgs("SELECT, INSERT",
+				user.AuthUsername, user.AuthHostname, tbl.Name)
+		}
```

Comment:
Why we don't just append items into `PlanBuilder.visitInfo` as other statements and check the privilege one for all in `Optimize`?

Reply:
Yes, I could do that.  In https://github.com/pingcap/tidb/pull/8426 I moved privilege checks out of `PlanBuilder.visitInfo` because I thought they were too contextual.  In this case, if I move the privilege checks first, analyzing multiple statements will at least be atomic.  As I understand, MySQL 8.0 is atomic for most of these commands, but MySQL 5.7 is not.

Reply:
Could you please explain more about the `atomic` part? I am a little bit confused, you mean using `visitInfo` to check privileges would make analyzing multiple statements atomic, which is not compatible with MySQL 5.7?

Reply:
My understanding of the behavior of MySQL 5.7: `ANALYZE TABLE t1, t2` will check permissions on t1, then analyze it, then check the permissions of t2, analyze it.  I haven't looked at `ANALYZE` specifically though.. but most work(ed) this way.  8.0 is usually atomic though.  One example I know is different is `DROP USER u1, u2`.

Reply:
I could of course also produce atomicity by looping through the tables twice in `buildAnalyzeTable`, but I will look at adding a check to `visitInfo`.